### PR TITLE
Focal Sample Reweighting: upweight hard-loss samples by loss^0.5

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The baseline's existing hard-sample mechanisms (PCGrad, tandem_ramp, pcgrad_extreme_pct) operate at the **gradient conflict** and **domain balance** levels — they don't dynamically prioritize samples that the model currently struggles with most. Adding focal-style per-sample reweighting at the **loss** level should direct more gradient signal to the hardest samples (extreme tandem configurations, unusual Re values) that consistently produce the highest surface error.

**Mechanism:** Compute the per-sample surface MAE, then weight each sample's contribution to the batch loss by `(loss_i / mean_loss)^gamma`. This is analogous to Focal Loss for classification (Lin et al., RetinaNet, CVPR 2017) applied to regression:
- gamma=0: standard uniform weighting (baseline)
- gamma=0.5: moderate upweighting of hard samples (2x error → 1.41x weight)
- gamma=1.0: aggressive upweighting (2x error → 2x weight)

**Key distinction from in-flight experiments:**
- **Frieren's OHNM (#2249):** reweights individual *nodes* within each sample based on per-node error. This reweights entire *samples* based on their aggregate surface MAE.
- **Nezuko's aft-foil upweight (#2253):** applies fixed 1.5x weight to aft-foil nodes. This is dynamic and sample-level.
- **PCGrad extreme pct:** only handles Re-extreme samples. This handles ANY hard sample.

Use gamma=0.5 as the first trial — moderate enough to not destabilize training but meaningful signal shift.

### 1. Add CLI flag to `Config` dataclass (near `pcgrad_extreme_pct`):
```python
focal_sample_gamma: float = 0.0  # focal-style per-sample loss reweighting; 0=disabled
```

### 2. Modify the per-sample loss computation

Find where the surface loss is averaged across the batch. The current structure computes surface loss per sample (or per node), then reduces with `.mean()`. You need to intercept just before the final `.mean()` reduction on surface loss.

The key change: **before the final `.mean()` call on surface_loss**, compute per-sample MAE and reweight:

```python
# After computing per-sample surface MAE (shape [B]):
if cfg.focal_sample_gamma > 0.0:
    # Compute mean loss across batch (detached to avoid second-order gradients)
    mean_loss = surface_loss_per_sample.detach().mean().clamp(min=1e-8)
    # Focal weights: samples with higher loss get more weight
    focal_weights = (surface_loss_per_sample.detach() / mean_loss).pow(cfg.focal_sample_gamma)
    # Normalize so total gradient magnitude is preserved
    focal_weights = focal_weights / focal_weights.mean()
    surface_loss = (surface_loss_per_sample * focal_weights).mean()
else:
    surface_loss = surface_loss_per_sample.mean()
```

**Implementation notes:**
- Apply focal reweighting to the SURFACE loss only, not the volume loss (we care about surface MAE)
- Use `.detach()` on the weights so no second-order gradients are computed
- Normalize weights to preserve gradient scale (prevents effective LR change)
- Apply AFTER the PCGrad split — modify each domain's surface loss separately within the PCGrad forward passes
- Log `focal_weights.std()` to W&B so we can see how much variance is in the weights

**Important:** You need to restructure the surface loss to give you per-sample values `[B]` before the `.mean()`. The current code may already compute this — look for where surface masks are applied and per-sample averaging is done.

### 3. Run configuration

```bash
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/focal-gamma05-s42" \
  --wandb_group "focal-sample-reweight" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --focal_sample_gamma 0.5 \
  --seed 42
```

Run with seeds 42 and 73. Use `--wandb_group "focal-sample-reweight"` for both runs.

### 4. Suggested follow-ups (for the comments section)
If gamma=0.5 works: try gamma=1.0 for stronger upweighting.
If gamma=0.5 hurts: the existing mechanisms already cover sample-level difficulty.

## Baseline (PR #2213, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in | 11.979 | < 11.98 |
| p_oodc | 7.643 | < 7.65 |
| p_tan | 28.341 | < 28.34 |
| p_re | 6.300 | < 6.30 |

Baseline W&B runs: hgml7i2r (s42), qic03vrg (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro --wandb_name "tanjiro/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Results — Focal Sample Reweighting (gamma=0.5)

**Implementation:** Per-sample focal-style loss reweighting applied to surface loss. Each sample weighted by `(loss_i / mean_loss)^0.5`, normalized to preserve gradient scale. Applied in main loss path, PCGrad 3-way groups, and SAM path.

**W&B runs:**
- Seed 42: [`3sjfouvz`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3sjfouvz) — 146 epochs
- Seed 73: [`jzcb7n89`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/jzcb7n89) — 147 epochs
- Group: `focal-sample-reweight`

### Surface Pressure MAE (lower is better)

| Metric | Baseline | Seed 42 | Seed 73 | Avg | vs Baseline |
|--------|----------|---------|---------|-----|-------------|
| p_in | 11.98 | 13.0 | 12.7 | 12.83 | **+7.1% ❌** |
| p_oodc | 7.65 | 8.3 | 7.9 | 8.10 | **+5.9% ❌** |
| p_tan | 28.34 | 28.4 | 28.7 | 28.55 | **+0.7% ❌** |
| p_re | 6.30 | 6.5 | 6.5 | 6.50 | **+3.2% ❌** |

### Analysis

All four metrics regressed. The focal reweighting at gamma=0.5 hurts performance across the board. The p_in regression is particularly severe (+7.1%).

**Why it likely failed:**
- The model already has multiple hard-sample mechanisms: PCGrad 3-way splits by domain difficulty, tandem_ramp adaptively boosts tandem samples, and asymmetric hard-node mining (epoch 30+) upweights hard nodes within samples. Adding focal sample reweighting on top creates over-correction — the hardest samples get boosted multiple times through different mechanisms.
- The `(loss/mean)^0.5` weights reduce the effective contribution of easy (well-predicted) samples, which likely include the in-distribution cases. This explains the large p_in regression.
- At gamma=0.5, a sample with 2x average loss gets 1.41x weight. Combined with tandem_boost (up to 4x) and hard-node mining (1.5x), the compounding effect is substantial.

**Conclusion:** The existing hard-sample mechanisms (PCGrad + tandem_ramp + hard-node mining) already cover sample-level difficulty sufficiently. Adding focal reweighting on top over-corrects and destabilizes the loss landscape.

### Suggested follow-ups
- This result suggests the current balance of hard-sample mechanisms is already near-optimal. Further gains are more likely from architecture/loss formulation changes than from additional sample weighting.
- If sample-level reweighting is revisited, it should REPLACE one of the existing mechanisms rather than stack on top.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2257 because ADVISOR: **Closing — clear dead end.** All 4 surface MAE metrics regressed significantly:

| Metric | Baseline | Focal γ=0.5 (2-seed) | Δ |
|--------|----------|----------------------|---|
| p_in | 11.979 | 12.83 | **+7.1%** |
| p_oodc | 7.643 | 8.10 | **+5.9%** |
| p_tan | 28.341 | 28.55 | **+0.7%** |
| p_re | 6.300 | 6.50 | **+3.2%** |

**Why:** The existing hard-sample mechanisms (PCGrad 3-way, tandem_ramp, hard-node mining) already cover sample-level difficulty. Stacking focal reweighting on top creates over-correction — the hardest samples get boosted multiple times through different mechanisms, and easy in-distribution samples lose too much gradient signal (hence the severe p_in regression).

**Takeaway:** The sample weighting stack is already near-optimal. Future gains are more likely from architecture, loss formulation, or new input features — not additional sample weighting. Good experiment, tanjiro — the negative result is informative and confirms the current balance of mechanisms is well-tuned..
